### ISSUE TITLE
After BLOB upload, the next reconnection caused by the SAS token timeout fails

### ISSUE THREAD

**Issue by TizianoFranzoi** [Azure/azure-iot-sdk-c#940]
**Environment:**
We use the azure C SDK 1.2.10, running in a linux environment on a ARM processor.

**Short Problem description:**
Once an upload to blob happened, the next reconnection caused by SAS token timeout fails. 

**Regular behavior:**
The application connects to an IoT hub and sends events from time to time. About all 45 minutes a short disconnect happens with reason IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN. Usually the connection is up again within about 1-2s. This is fairly stable. 

Some device methods are implemented and work well. One of them performs an upload to a blob storage. This upload works and finishes without an error.

**Error situation:**
But once this upload to blob upload is executed, the next disconnect caused by an IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN leads to a permanent loss of the connection. No reconnect happens anymore. The only method to get connected again seems to be to destroy the client using IoTHubClient_Destroy() and creating a new one. The new client will connect immediately.

This is reproducable.

Some relevant code fragments used:

Initialization and connect:
```
    IOTHUB_CLIENT_TRANSPORT_PROVIDER protocol = MQTT_Protocol;
  _iothub_client_handle = IoTHubClient_CreateFromConnectionString(
                  RP::instance()->di.connection_string.c_str(), protocol);
  bool traceOn = true;
  IoTHubClient_SetOption(_iothub_client_handle, OPTION_LOG_TRACE, &traceOn);
  auto retvalue = IoTHubClient_SetDeviceMethodCallback(_iothub_client_handle,device_method_callback, _parent);
            

Sending of events:            
  IoTHubClient_SendEventAsync(_iothub_client_handle, message_handle, 
                              send_confirm_callback,(void*)(uintptr_t) message_tracking_id);
            
Handling of device methods:            
int  iothub::impl::device_method_dispatcher(const std::string& method_name, 
                                            const std::string& payload, std::string& return_value)
  if (method_name = ....)
  ...
  else if (method_name =="LogDump") {
    IoTHubClient_UploadToBlobAsync(_iothub_client_handle, hub_file_name.c_str(), 
                                   (const unsigned char*)content.c_str(), size, 
                                    upload_callback,(void*)(uintptr_t) message_tracking_id);
```

**Comment by TizianoFranzoi** [comment]
I was able to catch azure sdk logs: after the SAS-Timeout the device is disconnected and the retries looks like following: 

```
2019-04-08 15:14:20.183 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:20.184 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:20.184 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:20.184 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:20.184 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address IHU-CSIoTHub-WEU-TST-01.azure-devices.net.
2019-04-08 15:14:21.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:21.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:21.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:21.002 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:21.003 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address IHU-CSIoTHub-WEU-TST-01.azure-devices.net.
2019-04-08 15:14:23.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:23.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:23.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:23.002 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:23.002 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address IHU-CSIoTHub-WEU-TST-01.azure-devices.net.


2019-04-08 15:14:27.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:27.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:27.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:27.001 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:27.002 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address IHU-CSIoTHub-WEU-TST-01.azure-devices.net.
2019-04-08 15:14:35.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:35.001 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:35.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:35.002 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:35.002 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address IHU-CSIoTHub-WEU-TST-01.azure-devices.net.
2019-04-08 15:14:51.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@428]Failed allocating OpenSSL context.
2019-04-08 15:14:51.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/log_ERR_get_error@435]  [0] error:140A90F1:lib(20):func(169):reason(241)
2019-04-08 15:14:51.002 ERROR [1448] [utils::azure_log@564] [tlsio_openssl.c/tlsio_openssl_open@1251]Failed creating the OpenSSL instance.
2019-04-08 15:14:51.002 ERROR [1448] [utils::azure_log@564] [mqtt_client.c/mqtt_client_connect@1001]Error: io_open failed
2019-04-08 15:14:51.003 ERROR [1448] [utils::azure_log@564] [iothubtransport_mqtt_common.c/SendMqttConnectMsg@2262]failure connecting to address `IHU-CSIoTHub-WEU-TST-01.azure-devices.net.``
```

**Comment by TizianoFranzoi** [comment]
We were able to solve the problem, but more by accident: 

After the IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN disconnect we call 
IoTHub_Deinit() followed by an IoTHubInit() call. 
No more ssl erros in the log and the device returns re-connected very fast again to the hub.. 
But we don't know of any impacts the DeInit/Init calls could have.

**Comment by ewertons** [comment]
Hi @TizianoFranzoi ,
thanks for sharing an update.

IoTHubInit() and IoTHub_Deinit() are supposed to be called just once at a time.

Calling these functions seems to be a mitigation of the issue, which is great because it unblocks you, but it seems to hide the real issue wherever it is.

Digging a little I found this post about [140A90F1:lib(20):func(169):reason(241)](https://marc.info/?l=openssl-users&m=122447060319674&w=2):

> The OpenSSL executable has the 'errstr' command for this purpose. The error indicates that the SSL v2 code could not load the MD5 routines.

I will have to dig a little more to provide more details.

Here is one more link, for our reference when we come back to this issue:
http://openssl.6102.n7.nabble.com/SSL-CTX-new-unable-to-load-ssl2-md5-routines-td20970.html

**Comment by TizianoFranzoi** [comment]
Hi @ewertons  thx for your additional  comments.
I agree completely that we only hide the real issue:

I want to add some more informations here: 

I did the same investigations on another device with  the same code but different Linux Environnement.
The Linux Kernel is newer, as the Gnu Compiler version.

The short result is, that we don't find any issue here.

**Comment by YoDaMa** [comment]
Hey @TizianoFranzoi, are you saying that the issue does not occur on a platform with a newer Linux Kernel, but is still reproducible on your initial platform?

**Comment by TizianoFranzoi** [comment]
Hey @YoDaMa. Yes the kernel is newer as the device is a newer one. But what is important to say is: Its another kernel, and its probably another version of the openssl lib. Or some other configs on the device..
our BSP-Vendor will have to investigate that.

**Comment by TizianoFranzoi** [comment]
Additional info: 
1. Device (the one with the issue): Kernel  3.4.53  / openssl 1.0.2p
2. Device (the one without problems): Kernel 4.4.68 / openssl 1.0.2h

**Comment by YoDaMa** [comment]
Thanks for the response @tizianofranzoi. Sounds like there’s specifically a bug with the client library using OpenSSL 1.0.2p, and possibly earlier. When we can get a repro on that version of OpenSSL we’ll investigate this bug. For the moment though it sounds like simply using the newer devices with the newer OpenSSL has mitigated this.

**Comment by YoDaMa** [comment]
@TizianoFranzoi one point that is confusing: You are saying that the Device without the problems has openssl 1.0.2h on it. Thats almost 2 years older than 1.0.2p.

>  OpenSSL_1_0_2h released on May 3, 2016
>  OpenSSL_1_0_2p released on Aug 14, 2018

Can we double check the versioning on the newer device with the newer Kernel version? Seems a bit off to have such an old release of OpenSSL on it.

**Comment by YoDaMa** [comment]
Hi @TizianoFranzoi it sounds like this bug is mitigated for your solution. I'm going to go ahead and close this thread, however if your BSP Vendor finds more information or finds additional questions that need addressing, please reopen this or a new issue.

**Comment by az-iot-builder-01** [comment]
@TizianoFranzoi, @ewertons, @YoDaMa, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by aog2000a** [comment]
Same issue with Azure IoT SDK for C, version 1.3.4, and OpenSSL 1.0.2s.

**Comment by aog2000a** [comment]
Hi @TizianoFranzoi, aren't you using either openssl directly, or some other library in your code which uses openssl (like libcurl)?

I think I've found the problem, at least in my case, and it is that my code uses OpenSSL, CURL and Azure IoT, and the initialization/initialization functions step on each other's toes with regard to OpenSSL...

Azure IoT tries to be clean by calling `curl_global_init(CURL_GLOBAL_NOTHING)`, but if some other code calls `curl_global_init(CURL_GLOBAL_SSL)' before Azure IoT calls `curl_global_cleanup()', then CURL ends up calling openssl's `EVP_cleanup()' which in practice deinitializes openssl... a new call to 'SSL_library_init();' is required.
